packages/lodestar/src/eth1/impl/ethers.ts

```diff
@@ -78,146 +76,95 @@ export class EthersEth1Notifier extends (EventEmitter as { new(): Eth1EventEmitt
       return;
     }
     this.started = true;
-    if(!this.contract) {
-      await this.initContract();
+    await this.initialize();
+  }
+
+  /**
+   * Main methods to be called either to build genesis state or to save data to propose blocks later.
+   * @param isScanEth1ForGenesis
+   * @param fromBlock
+   */
+  public async getDepositEventsFromBlock(
+    isScanEth1ForGenesis: boolean, fromBlock?: number): Promise<Pushable<IDepositEvent[]>> {
+    if (!isScanEth1ForGenesis && !this.opts.enabled) {
+      this.logger.warn(`eth1 is disabled so can't get deposit event, fromBlock=${fromBlock}`);
+      return null;
     }
-    const lastProcessedBlockTag = await this.getLastProcessedBlockTag();
-    this.lastProcessedEth1BlockNumber = (await this.getBlock(lastProcessedBlockTag)).number;
-    this.lastProcessedDepositIndex = await this.getLastProcessedDepositIndex();
-    this.logger.info(
-      `Started listening to eth1 provider ${this.opts.provider.url} on chain ${this.opts.provider.network}`
-    );
-    this.logger.verbose(
-      `Last processed block number: ${this.lastProcessedEth1BlockNumber}, ` +
-      `last processed deposit index: ${this.lastProcessedDepositIndex}`
-    );
-    const headBlockNumber = await this.provider.getBlockNumber();
-    // process historical unprocessed blocks up to curent head
-    // then start listening for incoming blocks
-    this.processBlocks(headBlockNumber).then(() => {
-      if(this.started) {
-        this.provider.on("block", this.onNewEth1Block.bind(this));
-      }
-    });
+    // if we try to form genesis, allow to scan deposit events even eth1 is disabled
+    if (!this.opts.enabled) {
+      await this.initialize();
+    }
+    this.lastProcessedEth1BlockNumber = fromBlock || this.opts.depositContract.deployedAt - 1;
+    this.logger.info("eth1: Get deposit events by block from block " + this.lastProcessedEth1BlockNumber);
+    this.eth1DataSource = pushable<IDepositEvent[]>();
+    this.targetBlockSource = pushable<number>();
+    // should not wait
+    this.doGetDepositEventsFromLastProcessedBlock();
+    return this.eth1DataSource;
+  }
+
+  /**
+   * Called by genesis builder to inform that we found genesis state.
+   * Should stop pushing deposit events to genesis builder.
+   */
+  public async foundGenesis(): Promise<void> {
```

Comment:
Instead of this method, it's better to solve it using abort signal as param, which will trigger source closing